text/0000-pub-sub.md

```diff
@@ -0,0 +1,171 @@
+
+#### Event delivery
+There is a high-level sequence diagram below which represent the event delivery flow.
+
+![Event delivery diagram](http://www.plantuml.com/plantuml/png/nLPDJnj13Btlhx32WJQ22AweHOgWQYjH3f0RuJ3PSR8Z9fxjp6peIldnzUntabL80r8Fa6JNVlPxzjZZbTS8trP87YHgjKARGsfDZCufMaIHoBoXR3b32uCli-kdYukBvq4KdQR2UXMhLB3ZO6da-PSqcsdb4hJN0n0EqYYAZTOOUsN8bJPiuh9PyHebHj4k2BzsiLKSxdnSuhZBlMQZqQYoa_lXxYipwWkxctj4nYCOrgD85M0u76iDBQ03Yp6oO811FKPMoqnxGMWof_FXC4AI452YzNeTCGYSNmUiAtX0RnLxWQ3I6URMBC4d21b9j3fdwQV4PLWf_daQGZcJsHZXapUfYi_WTuOPdW7wUFHvIcFj3CGcqn9c27DbaR_aMu1GustayeYXH20skKtfDL5n0i9oabp1bV0CEJSM17Wa5cvAuPTFbAie-3pbYXRGJoVtOfK_Ln76KbOcsYo0FstERQKzJ5XlB4AGwGPdiG6wETua6B-K6Ap5P79NM0I8id4v1zOPy4yQc1Sljo4UqCK9oanNvMgQdPagpqJ30v5l-hxDep3L3jj4jfDOM9EbC7OvnSsp9W1dbrx1lV5gddCAJHT25WvVqHxMiimnbsScM8EgX7BK9ahzX6-vhjp0eGMAf8zYnjW5kxPItVxlRuoePLqUcP2y0sysVHahHSkEI_RVEJfridowDn8tdFbtYqZFpDDT5ZH36LkU8P3eX6hSwspP3KvkXNrv57ffg7dFZxlWXJ2J1ZNXgJ3ZL_MXbyVekQyVGlYIqWz1KdSgnmE9ZHDhg52dj_8IE_GEL4yUfvxiLq_-N_KaL9JUgvxiLM-_adISajIl9465kKrbm-GVnEOv8kkTZlsiwTZni2tGxjX-JT27aIRO7ONpOAWxHgOKycrItIHRROd3kkWZ1ryrz3g1mu9vEtZ_fZfeMxrtOnrnclPlhdtdxICts7Oh7LpGOnXNSS9dLFpJkAWVhh_2dUBcvCy9NsPWNJxbIaWOCujN_lwJF5eIzLQhJKBynkQNKEJvhn00l1LAWvb3UVajpx_o2bllM-F35kt6kQdrAiP8AsjOe-5UsGYxXH66GWEoTQqFOw0zU7nV9BufBaMudpDCGAt4YFu2)
+
+#### Security
+Calls to the `pubsub` can be done either on behalf of a user or a backend module if this is a result of the processing of another event. While all calls from `mod-pubsub` to subscribers will be done on behalf of the `pubsub`. For these cases we can’t use or apply permissions granted to a user, because these calls don't have user context and could be treated as “System” or “Internal”. Since FOLIO Platform currently does not provide such type of users (a System/Internal level user) it makes sense to create an ordinary user that will be assigned to the `pubsub` to perform all actions on behalf of that. Also, it means that all data required by OKAPI to be provided in the HTTP headers must be transferred as attributes in the metadata section of an event payload (X-Okapi-Tenant, etc.). The drawback of this approach is that the dedicated user can be accidentally deleted by a FOLIO administrator. Also, it is allowed to login from UI using that user’s credentials.
+
+The `mod-pubsub` module should be able to obtain a JWT token from mod-authtoken on behalf of that “technical” user as well. It could be done when the module is enabled for a tenant. Time-to-live for such a token should be quite long enough.
```

Comment:
Will the request to the `delivery endpoint` be made via Okapi? 

And hence, a new JWT token would be generated with the combined permissions of the `system user` and the `module permissions` declared for that endpoint?

If so, that would mean each delivery endpoint would need to be declared in the module descriptor of the subscribing module. That might mean we need a marker interface similar to `_timer` defined for scheduler invoked endpoints. 

If not, then the `system user` would likely need the super set of all permissions needed by each subscriber endpoint to fulfil their activity.

Reply:
What the proposal deliberately left unspecified is that there is no constraint on how the subscribing module provide the delivery endpoints. It can specify a single endpoint "handler" for all the event types it want to subscribe to. Or it may choose to specify a finite set of them, precisely so as to be able to control permissions.

Since this would likely be done at tenant activation time, it is not inconceivable to simply have a set of desired event types and their required delivery permissions, then bulk assign them to the "system" use ("tenant" user really) programmatically.

I think the difference with the _timer case, is that here the delivery originate outside of Okapi: in mod-pubsub which is a module. That module has the ability to store on a tenant basis the credentials that it will use to originate deliveries.

Reply:
> What the proposal deliberately left unspecified is that there is no constraint on how the subscribing module provide the delivery endpoints. It can specify a single endpoint "handler" for all the event types it want to subscribe to. Or it may choose to specify a finite set of them, precisely so as to be able to control permissions.

Sure, there could be one or many receiving endpoints at the discretion of the subscriber.

Ah, I think I may have poorly asked my question. Maybe I’m asking the wrong question. And your answer has helped guide that, thank you.

Are delivery endpoints defined as paths, and mod-pubsub resolves them via Okapi, or are they opaque endpoints, and it is entirely up to the subscriber?

Based upon your answer here, it seems it is the latter. The receiver could use an Okapi based address, and hence use facilities like module permissions if they want to, or it could choose not to.

> Since this would likely be done at tenant activation time, it is not inconceivable to simply have a set of desired event types and their required delivery permissions, then bulk assign them to the "system" use ("tenant" user really) programmatically.

I agree that is not inconceivably.

Is that what we are proposing to do, or is this aspect out of scope for this RFC or deliberately left up to the subscriber?

> I think the difference with the _timer case, is that here the delivery originate outside of Okapi: in mod-pubsub which is a module. That module has the ability to store on a tenant basis the credentials that it will use to originate deliveries.

I wasn’t intending to suggest that this was the same as the _timer interface for scheduled tasks. Apologies for that confusion.

Reply:
> Are delivery endpoints defined as paths, and mod-pubsub resolves them via Okapi, or are they opaque endpoints, and it is entirely up to the subscriber?

It is just another module, so endpoints are paths and mod-pubsub resolves them via Okapi.

Reply:
Thanks for clarifying that expectation of a subscriber. It means that delivery endpoints have to be defined in a subscriber's module descriptor (and those endpoints can use facilities like module permissions)